### DISCUSSION TITLE
Berry | Driver does not subscribe to MQTT topic on restart

### DISCUSSION THREAD

**Discussion by RPJacobs** [arendst/Tasmota#16460] [Q&A]
HI,

I've created a i2c driver to controle a sauna heater. 

In the init function i subscribe to two topics. This works if i run it directly in the berry console. But on boot (autoexec.be) the driver works but does not subscribe to mqtt topics.

Please advice

```
  def init()
    self.temp = 0
    self.hum = 0
    self.count = 0
    self.setpoint = 90
    self.power = "OFF"

    self.wire = tasmota.wire_scan(0x40)

    if self.wire
      print("I2C: Sensor detected on bus "+str(self.wire.bus))
    end
    self.subscribe()
    tasmota.add_driver(self)
  end
  
  def subscribe()
    mqtt.subscribe("dummy/saunaSetpoint")
    mqtt.subscribe("dummy/saunaPower")
  end

  #- get mqtt values -#
  def mqtt_data(topic, idx, payload_s, payload_b)
      if (topic == "dummy/saunaSetpoint")
	      self.setpoint = int(payload_s)
	       return true
      end		
      if (topic == "dummy/saunaPower")
	      self.power = payload_s
	       return true
      end
	      return false
  end
```

**Comment by sfromis** [comment]
At `autoexec.be` time, there is no MQTT connection yet. You need to delay subscribe till after that. Possibly by having a trigger on `System#Boot` to get in when Tasmota is fully initialized. Beware that things like temporary outages of Wifi or MQTT may cause loss of subscriptions too, you basically need to manage subscriptions.

In your case, instead of using MQTT subscribe at all, I'd suggest registering Tasmota commands via `tasmota.add_cmd` to get the setpoint and power values in a simpler and more reliable way.

Assuming still using MQTT subscribe for whatever other reason, this is a much shorter way of doing it:
```
  def subscribe()
    mqtt.subscribe("dummy/saunaSetpoint", def (topic,idx,payload_s,payload_b) self.setpoint=int(payload_s) return true end)
    mqtt.subscribe("dummy/saunaPower", def (topic,idx,payload_s,payload_b) self.power=int(payload_s) return true end)
  end
```
You need no `mqtt_data` then, when using small anonymous functions that way.

**Reply by RPJacobs** [reply]
Thanks for your reply. MQTT not being connected in autoexec makes sense. 

Is there a way to get System#Boot in berry? Or MQTT#connected? 

I will check the manual on Tasmota.add_cmd and subscribing to a MQTT topic.

**Reply by sfromis** [reply]
I only wrote about using `System#Boot` (and so on, including `Mqtt#Connected`) because it is available in Berry, using the regular `tasmota.add_rule`. Of course, if you go for my suggested simpler strategy of registering commands, you need not care about MQTT subscription timing, or anything MQTT at all, as Tasmota obviously automatically manages subscriptions on the command topic.

**Reply by RPJacobs** [reply]
Hmmm setpoint en power are values set by openhab on MQTT. So I will have to use MQTT.

**Reply by sfromis** [reply]
If you have openhab publish to the Tasmota command topic, you are using MQTT, but you do not need any MQTT handling in Berry. That was the point of using commands (via MQTT) instead of the issues having good subscription management in Berry. Should be as easy in openhab, just a different topic like `cmnd/tasmotadevice/setpoint` with the payload.

**Reply by RPJacobs** [reply]
I fixed it by adding "tasmota.add_rule("mqtt#connected", /->self.subscribe())" in the init function. Thanks!

**Reply by sfromis** [reply]
With most recent dev Tasmota, the Berry subscription management has been enhanced to manage subscriptions across MQTT dis/connects, meaning that your rule there should now be superfluous, and you can go back to simply having the subscribe directly in `autoexec.be`.

I'd still much recommend using commands instead of a MQTT subscription, just better for the purpose.
